src/openrct2-ui/windows/Viewport.cpp

```diff
@@ -10,192 +10,220 @@
-    rct_window* mainWindow = window_get_main();
-    if (mainWindow != nullptr)
+private:
+    void GetFreeViewportNumber()
     {
-        rct_viewport* mainViewport = mainWindow->viewport;
-        int32_t x = mainViewport->viewPos.x + (mainViewport->view_width / 2);
-        int32_t y = mainViewport->viewPos.y + (mainViewport->view_height / 2);
-        w->savedViewPos = { x - (w->viewport->view_width / 2), y - (w->viewport->view_height / 2) };
+        number = 1;
+        window_visit_each([&](rct_window* w) {
+            if (w != nullptr && w != this && w->classification == WC_VIEWPORT)
+            {
+                if (w->number >= number)
+                    number = w->number + 1;
+            }
+        });
     }
 
-    w->viewport->flags |= VIEWPORT_FLAG_SOUND_ON;
-
-    return w;
-}
+public:
+    void OnOpen() override
+    {
+        GetFreeViewportNumber();
+
+        widgets = window_viewport_widgets;
+        enabled_widgets = (1ULL << WIDX_CLOSE) | (1ULL << WIDX_ZOOM_IN) | (1ULL << WIDX_ZOOM_OUT) | (1ULL << WIDX_LOCATE);
+
+        // Create viewport
+        viewport_create(this, windowPos, width, height, 0, TileCoordsXYZ(128, 128, 0).ToCoordsXYZ(), 1, SPRITE_INDEX_NULL);
+        if (viewport == nullptr)
+        {
+            Close();
+            window_error_open("Unexpected Error", "Failed to create viewport window.");
+            return;
+        }
+
+        auto* mainWindow = window_get_main();
+        if (mainWindow != nullptr)
+        {
+            rct_viewport* mainViewport = mainWindow->viewport;
+            int32_t x = mainViewport->viewPos.x + (mainViewport->view_width / 2);
+            int32_t y = mainViewport->viewPos.y + (mainViewport->view_height / 2);
+            savedViewPos = { x - (viewport->view_width / 2), y - (viewport->view_height / 2) };
+        }
+
+        viewport->flags |= VIEWPORT_FLAG_SOUND_ON;
+
+        min_width = WW;
+        min_height = WH;
+        max_width = WW;
+        max_height = WH;
+    }
 
-static void window_viewport_anchor_border_widgets(rct_window* w)
-{
-    w->widgets[WIDX_BACKGROUND].right = w->width - 1;
-    w->widgets[WIDX_BACKGROUND].bottom = w->height - 1;
-    w->widgets[WIDX_PAGE_BACKGROUND].right = w->width - 1;
-    w->widgets[WIDX_PAGE_BACKGROUND].bottom = w->height - 1;
-    w->widgets[WIDX_TITLE].right = w->width - 2;
-    w->widgets[WIDX_CLOSE].left = w->width - 13;
-    w->widgets[WIDX_CLOSE].right = w->width - 3;
-}
+    void OnUpdate() override
+    {
+        auto* mainWindow = window_get_main();
+        if (mainWindow == nullptr)
+            return;
+
+        if (viewport != nullptr && viewport->flags != mainWindow->viewport->flags)
+        {
+            viewport->flags = mainWindow->viewport->flags;
+            Invalidate();
+        }
+
+        // Not sure how to invalidate part of the viewport that has changed, this will have to do for now
```

Comment:
What is this comment for? If it's about actual changes within the viewport you don't need to do anything that's already accounted for

Reply:
That was left from the original, before the refactor. I left it there in case the note is still relevant.